csrc/kernel_cache.cpp

```diff
@@ -779,10 +779,31 @@ FusionKernelRuntime* FusionExecutorCache::getKernelRuntimeFor(
   // that whenever we encounter a new set of input shapes we segment and compile
   // a new FusionKernelRuntime.
   if (!isOptionDisabled(DisableOption::KernelReuse)) {
+    std::optional<bool> should_not_be_segmented = std::nullopt;
     auto reuse_it = std::find_if(
         kernel_runtimes.begin(),
         kernel_runtimes.end(),
-        [&args, &new_heuristics, &forced_index_type](auto& kernel_runtime) {
+        [&should_not_be_segmented, &args, &new_heuristics, &forced_index_type](
+            auto& kernel_runtime) {
+          if (!should_not_be_segmented.has_value()) {
+            // Use the concretized complete fusion from this runtime to check
+            // whether we can schedule the complete fusion without segmenting.
+            std::unique_ptr<Fusion> conc_fusion = std::make_unique<Fusion>(
+                *(kernel_runtime->fusionSegments()->completeFusion()));
+            SchedulerRuntimeInfo runtime_info(
+                conc_fusion.get(),
+                args,
+                /*precomputed_values=*/nullptr,
+                /*all_tvs=*/{},
+                forced_index_type);
+            should_not_be_segmented = SchedulerEntry::proposeHeuristics(
```

Comment:
`SchedulerEntry::proposeHeuristics` returns a `std::optional<ScheduleHeuristic>` , can we assign it to `std::optional<bool>`?

Reply:
Currently I'm just checking that it has a value, and the actual heuristic will be found again below when defining `maybe_heuristics` on an unsegmented `FusionKernelRuntime`.